# Abundance analysis of metagenomes

This snakemake workflow can be used to analyse the abundance of species groups, based on different proteins. The metagenomes to study are defined in the [samples.tsv](./config/samples.tsv) file and the proteins to use as markers are defined in the [proteins.tsv](./config/proteins.tsv) file.

The columns to use in the [samples.tsv](./config/samples.tsv) file are:
- **sample_name**: name of sample
- **scaffolds_path**: path to scaffolds or contigs
- **contig_to_bin**: path to the contig to bin file (generated by DAStool)
- **forward_read_path**: path toforward reads
- **reverse_read_path**: path toreverse reads

The columns to use in the [proteins.tsv](./config/proteins.tsv) file are:

- **protein_name**: name of protein
- **hmm_path**: path to hmm file
- **filter_length_nt**: Remove protein hits with shorter length than specified.
- **cluster_threshold**: Currently not used, uses default of 99% (should use thresholds from [Matt olm et al.](https://journals.asm.org/doi/full/10.1128/mSystems.00731-19))


## Test command to run

```snakemake -nrp --cluster-config config/cluster.yaml --default-resources partition=standard --cluster "sbatch -J WYRPclust -p {cluster.p} -o {cluster.o}" -j10  --rerun-incomplete --latency-wait 60 --cluster-cancel scancel > test_run_p.txt```

S3 abundance of bins:
![](./results/S3c_phylum_rel_abund_uncl.svg)

Gene presence based on S3 abundance of bins:
![](./results/phylum_rmlA,rmlB,wzc,wzy,pelA,epsC,betB,dnaK,otsA,ompR,proA,mscL,osmC,nasA,xoxF,coxL,moaA,cynS,nthA,pmoB_3depths.svg)

